clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java

```diff
@@ -210,6 +220,38 @@ public void testMultipleServerMechanisms() throws Exception {
     }
 
     /**
+     * Tests that Kafka ApiVersionRequests are handled by the SASL server authenticator
+     * prior to SASL handshake flow and that subsequent authentication succeeds.
+     */
+    @Test
+    public void testUnauthenticatedApiVersionRequest() throws Exception {
+        SecurityProtocol securityProtocol = SecurityProtocol.SASL_PLAINTEXT;
+        configureMechanisms("PLAIN", Arrays.asList("PLAIN"));
+        server = NetworkTestUtils.createEchoServer(securityProtocol, saslServerConfigs);
+
+        // Create PLAINTEXT connection to manually authenticate after ApiVersionsRequest
+        String node = "1";
+        createClientConnection(SecurityProtocol.PLAINTEXT, node);
```

Comment:
It isn't about blocking vs non-blocking.
It is about how a Kafka broker behaves to a non-authenticated request on SASL ports... we need to test it (since it is the requirement of this JIRA).

We hope that SASL behaves exactly the same as PLAINTEXT, but we don't know that it does without a test.